presto-main/src/main/java/com/facebook/presto/operator/PageSupplier.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.Page;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import javax.annotation.Nullable;
+
+public interface PageSupplier
```

Comment:
Maybe this should be a more generic concept of iterator, for instance:
```
public interface BlockingIterator<T> extends Iterator<T> {
  // equivalent of isFinished (false when no more elements are expected)
  boolean hasNext();
  // returns next element, might block
  @Nullable
  T next();
  ListenableFuture<?> isBlocked();
}
```
It would be then possible to cleanly extend `MergeHashSort` and use it in distributed sort code. Also the abstraction could be used in other places, e.g: `MergeSource` and `MergeSortProcessor` could implement `BlockingIterator<PageWithPosition>` and `BlockingIterator<Page>`. One could even imagine standard transformation methods and base classes for such iterators (similarly as for futures and Guava `AbstractIterator`). There could be a utility method that would decorate existing iterator with blocking future.